app/lib/actions/katello/content_view/incremental_updates.rb

```diff
@@ -29,14 +35,38 @@ def plan(version_environments, composite_version_environments, content, dep_solv
             end
 
             if hosts.any? && !content[:errata_ids].blank?
-              errata = ::Katello::Erratum.with_identifiers(content[:errata_ids])
-              hosts = hosts.where(:id => ::Katello::Host::ContentFacet.with_applicable_errata(errata).pluck(:host_id))
               plan_action(::Actions::BulkAction, ::Actions::Katello::Host::Erratum::ApplicableErrataInstall, hosts, content[:errata_ids])
             end
             plan_self(:version_outputs => output_for_version_ids)
           end
         end
 
+        def find_applicable_hosts_and_smart_proxies(hosts, content)
+          smart_proxies = []
+          if hosts.any? && !content[:errata_ids].blank?
```

Comment:
Here's a possible refactor of this block of code:

```
-          if hosts.any? && !content[:errata_ids].blank?
+          if hosts.any? && content[:errata_ids].present?
             errata = ::Katello::Erratum.with_identifiers(content[:errata_ids])
-            hosts = hosts.where(:id => ::Katello::Host::ContentFacet.with_applicable_errata(errata).pluck(:host_id))
-            smart_proxies = Katello::Host::SubscriptionFacet.where(:host => hosts).pluck(:registered_through).uniq.compact
-            smart_proxies = SmartProxy.where(:name => smart_proxies).where.not(:id => SmartProxy.default_capsule!.id)
+            facets = ::Katello::Host::ContentFacet.with_applicable_errata(errata)
+            hosts = hosts.where(id: facets.pluck(:host_id))
+            smart_proxies = SmartProxy.non_default.where(content_facets: facets)
```
```
+++ b/app/models/katello/concerns/smart_proxy_extensions.rb
@@ -55,6 +55,7 @@ module Katello
           :message => _("must be one of the following: %s") % DOWNLOAD_POLICIES.join(', ')
         }
         scope :with_content, -> { with_features(PULP_FEATURE, PULP_NODE_FEATURE) }
+        scope :non_default, -> { with_features(PULP_NODE_FEATURE) }
```

I haven't tested it yet but it should solve the problem of looking up the proxies by name

Reply:
Sounds nice, will improve this way.